### ISSUE TITLE
NO audio I/O Backend is avaliable

### ISSUE THREAD

**Issue by ElinLiu0** [dotnet/TorchSharp#1453]
Hi:
  I'm currently want to remake a Voice changing model called [RVC](https://github.com/RVC-Project/Retrieval-based-Voice-Conversion-WebUI).
  Yet when i want to use `torchaudio.load()` method to load audio with below code:
```c#
public static NDarray loadAudio16k(string name)
{
    (torch.Tensor,int) audio40kTensor = torchaudio.load(name);
    return np.array((float[])audio40kTensor.Item1.tolist(), dtype: np.float32);
}
```

It raised below exception:
```c#
System.InvalidOperationException:“No audio I/O backend is available.
```

It most likely jumped to [this](https://github.com/dotnet/TorchSharp/blob/5ba14022749a20b98798105a32988da59f2356ca/src/TorchAudio/NoBackend.cs#L20) line.

I've installed NAudio package via Nuget on my project,why it happend?

**Comment by ElinLiu0** [comment]
> I would be great to list what needed to make TorchSharp Audio more minimum complete 

That would be so great if you convinced.

**Comment by GeorgeS2019** [comment]
https://github.com/dotnet/TorchSharp/discussions/598

**Comment by GeorgeS2019** [comment]
https://pytorch.org/audio/stable/tutorials/audio_io_tutorial.html

Many functions I believe there's are c# codes

**Comment by yueyinqiu** [comment]
I remember I have tried this with NAudio, but found some problems.

What I could remember is that NAudio support some formats with Windows Media Player, so no cross-platform support for those formats, and there are also some bugs on audio length.

This is a package I wrote for my own use to load audio as tensors: https://github.com/yueyinqiu/YueYinqiu.Su.TorchSharpUtilities/blob/main/src/TorchSharpUtilities/Extensions/WavFileAccessExtensions.cs

The code starting from line 51 is commented due to the audio length bug.

**Comment by GeorgeS2019** [comment]
Pytorch audio used ffmepeg

**Comment by ElinLiu0** [comment]
> I remember I have tried this with NAudio, but found some problems.
> 
> What I could remember is that NAudio supports some formats with Windows Media Player, so no cross-platform support for them, and there are also some bugs on audio length.
> 
> This is a package I wrote for my own use to access audios as tensors: https://github.com/yueyinqiu/YueYinqiu.Su.TorchSharpUtilities/blob/main/src/TorchSharpUtilities/Extensions/WavFileAccessExtensions.cs
> 
> The code starting from line 51 is commented due to the audio length bug. (might be another bug, I'm not sure if I remember it correctly)

我已经试着让DeepSeek-R1帮我写一版了，然后不知道是精度炸了还是啥，整个数据读出来和Librosa的实现完全就是一个北边儿一个南边儿 : (

**Comment by yueyinqiu** [comment]
@ElinLiu0 可能是 librosa 有归一化什么的？试试我的实现，或者画个图像看看呢？

**Comment by ElinLiu0** [comment]
> @ElinLiu0 可能是 librosa 有归一化什么的？试试我的实现，或者画个图像看看呢？

用Matplotlib 画了，频宽不一样

**Comment by yueyinqiu** [comment]
![Image](https://github.com/user-attachments/assets/c4076909-c3a7-4a97-8d09-f67dda9ab631)

我跑出来好像是一样的欸？

**Comment by ElinLiu0** [comment]
> ![Image](https://github.com/user-attachments/assets/c4076909-c3a7-4a97-8d09-f67dda9ab631)
> 
> 我跑出来好像是一样的欸？会不会是选的采样率不对？

是你的实现吗？我用的是 40000 和 16000

**Comment by yueyinqiu** [comment]
素的，用的是我的实现

Csharp:

```csharp
using NAudio.Wave;
using TorchSharp;
using YueYinqiu.Su.TorchSharpUtilities.Extensions;

using var audio = new AudioFileReader("C:\\Users\\yueyi\\Desktop\\actual.wav");

var tensor = audio.ReadAsTensor();
Console.WriteLine(tensor.cstr());
```

```csharp
    <PackageReference Include="TorchSharp-cpu" Version="0.105.0" />
    <PackageReference Include="YueYinqiu.Su.TorchSharpUtilities" Version="0.0.21" />
```

Python:

```python
from inputs_and_outputs.tensor_audios.tensor_audios import TensorAudios


py = TensorAudios.load_audio(r"C:\Users\yueyi\Desktop\actual.wav", 16000, "first_only")
print(py)
```

```python
from pathlib import Path
from typing import Literal

import numpy
from statictorch import Tensor1d, Tensor2d
from torch import Tensor

from basic_utilities.static_class import StaticClass


class TensorAudios(StaticClass):
    @classmethod
    def load_audio(cls,
                   path: Path,
                   sample_rate: int,
                   mutichannel_behavior: Literal["first_only", "as_mono", "as_many"]) \
                    -> Tensor2d:
        import librosa
        import numpy
        import torch

        numpy_: numpy.ndarray
        numpy_, _ = librosa.load(path, 
                                 sr=sample_rate, 
                                 mono=mutichannel_behavior == "as_mono")
        result: Tensor = torch.tensor(data=numpy_, dtype=torch.float)

        if result.shape.__len__() == 1:
            result = result.unsqueeze(0)
        elif mutichannel_behavior == "first_only":
            result = result[0:1, :]
        
        return Tensor2d(result)

    @classmethod
    def save_audio(cls, 
                   audio: Tensor1d | Tensor2d, 
                   path: Path, 
                   sample_rate: int) -> None:
        audio_numpy: numpy.ndarray = audio.detach().cpu().numpy()
        if audio_numpy.shape.__len__() == 2 and audio_numpy.shape[0] == 1:
            audio_numpy = audio_numpy.squeeze(0)
        import soundfile
        soundfile.write(path, audio_numpy, sample_rate)
```

**Comment by ElinLiu0** [comment]
> 素的，用的是我的实现
> 
> Csharp:
> 
> ```csharp
> using NAudio.Wave;
> using TorchSharp;
> using YueYinqiu.Su.TorchSharpUtilities.Extensions;
> 
> using var audio = new AudioFileReader("C:\\Users\\yueyi\\Desktop\\actual.wav");
> 
> var tensor = audio.ReadAsTensor();
> Console.WriteLine(tensor.cstr());
> ```
> 
> ```csharp
>     <PackageReference Include="TorchSharp-cpu" Version="0.105.0" />
>     <PackageReference Include="YueYinqiu.Su.TorchSharpUtilities" Version="0.0.21" />
> ```
> 
> Python:
> 
> ```python
> from inputs_and_outputs.tensor_audios.tensor_audios import TensorAudios
> 
> 
> py = TensorAudios.load_audio(r"C:\Users\yueyi\Desktop\actual.wav", 16000, "first_only")
> print(py)
> ```
> 
> ```python
> from pathlib import Path
> from typing import Literal
> 
> import numpy
> from statictorch import Tensor1d, Tensor2d
> from torch import Tensor
> 
> from basic_utilities.static_class import StaticClass
> 
> 
> class TensorAudios(StaticClass):
>     @classmethod
>     def load_audio(cls,
>                    path: Path,
>                    sample_rate: int,
>                    mutichannel_behavior: Literal["first_only", "as_mono", "as_many"]) \
>                     -> Tensor2d:
>         import librosa
>         import numpy
>         import torch
> 
>         numpy_: numpy.ndarray
>         numpy_, _ = librosa.load(path, 
>                                  sr=sample_rate, 
>                                  mono=mutichannel_behavior == "as_mono")
>         result: Tensor = torch.tensor(data=numpy_, dtype=torch.float)
> 
>         if result.shape.__len__() == 1:
>             result = result.unsqueeze(0)
>         elif mutichannel_behavior == "first_only":
>             result = result[0:1, :]
>         
>         return Tensor2d(result)
> 
>     @classmethod
>     def save_audio(cls, 
>                    audio: Tensor1d | Tensor2d, 
>                    path: Path, 
>                    sample_rate: int) -> None:
>         audio_numpy: numpy.ndarray = audio.detach().cpu().numpy()
>         if audio_numpy.shape.__len__() == 2 and audio_numpy.shape[0] == 1:
>             audio_numpy = audio_numpy.squeeze(0)
>         import soundfile
>         soundfile.write(path, audio_numpy, sample_rate)
> ```

好的，我回家之后试试

**Comment by ElinLiu0** [comment]
> 素的，用的是我的实现
> 
> Csharp:
> 
> using NAudio.Wave;
> using TorchSharp;
> using YueYinqiu.Su.TorchSharpUtilities.Extensions;
> 
> using var audio = new AudioFileReader("C:\\Users\\yueyi\\Desktop\\actual.wav");
> 
> var tensor = audio.ReadAsTensor();
> Console.WriteLine(tensor.cstr());
>     <PackageReference Include="TorchSharp-cpu" Version="0.105.0" />
>     <PackageReference Include="YueYinqiu.Su.TorchSharpUtilities" Version="0.0.21" />
> Python:
> 
> from inputs_and_outputs.tensor_audios.tensor_audios import TensorAudios
> 
> 
> py = TensorAudios.load_audio(r"C:\Users\yueyi\Desktop\actual.wav", 16000, "first_only")
> print(py)
> from pathlib import Path
> from typing import Literal
> 
> import numpy
> from statictorch import Tensor1d, Tensor2d
> from torch import Tensor
> 
> from basic_utilities.static_class import StaticClass
> 
> 
> class TensorAudios(StaticClass):
>     @classmethod
>     def load_audio(cls,
>                    path: Path,
>                    sample_rate: int,
>                    mutichannel_behavior: Literal["first_only", "as_mono", "as_many"]) \
>                     -> Tensor2d:
>         import librosa
>         import numpy
>         import torch
> 
>         numpy_: numpy.ndarray
>         numpy_, _ = librosa.load(path, 
>                                  sr=sample_rate, 
>                                  mono=mutichannel_behavior == "as_mono")
>         result: Tensor = torch.tensor(data=numpy_, dtype=torch.float)
> 
>         if result.shape.__len__() == 1:
>             result = result.unsqueeze(0)
>         elif mutichannel_behavior == "first_only":
>             result = result[0:1, :]
>         
>         return Tensor2d(result)
> 
>     @classmethod
>     def save_audio(cls, 
>                    audio: Tensor1d | Tensor2d, 
>                    path: Path, 
>                    sample_rate: int) -> None:
>         audio_numpy: numpy.ndarray = audio.detach().cpu().numpy()
>         if audio_numpy.shape.__len__() == 2 and audio_numpy.shape[0] == 1:
>             audio_numpy = audio_numpy.squeeze(0)
>         import soundfile
>         soundfile.write(path, audio_numpy, sample_rate)

哥们儿能有空来看看这个是为啥报错了吗？https://github.com/dotnet/TorchSharp/issues/1456

**Comment by yueyinqiu** [comment]
#1377
